Librerie/libpdel-0.6.1/pdel/net/route_msg.3
.\" @COPYRIGHT@
.\"
.\" Author: Archie Cobbs <archie@freebsd.org>
.\"
.\" $Id: route_msg.3 901 2004-06-02 17:24:39Z archie $
.\"
.Dd April 22, 2002
.Dt ROUTE_MSG 3
.Os
.Sh NAME
.Nm route_msg
.Nd routing socket message objects
.Sh LIBRARY
PDEL Library (libpdel, \-lpdel)
.Sh SYNOPSIS
.In sys/types.h
.In stdio.h
.In pdel/net/route_msg.h
.Ft "struct route_msg *"
.Fn route_msg_create "void"
.Ft "void"
.Fn route_msg_destroy "struct route_msg **msgp"
.Ft "int"
.Fn route_msg_get_type "struct route_msg *msg"
.Ft "void"
.Fn route_msg_set_type "struct route_msg *msg" "int type"
.Ft "int"
.Fn route_msg_get_index "struct route_msg *msg"
.Ft "void"
.Fn route_msg_set_index "struct route_msg *msg" "int index"
.Ft "int"
.Fn route_msg_get_flags "struct route_msg *msg"
.Ft "void"
.Fn route_msg_set_flags "struct route_msg *msg" "int flags"
.Ft "int"
.Fn route_msg_get_error "struct route_msg *msg"
.Ft "pid_t"
.Fn route_msg_get_pid "struct route_msg *msg"
.Ft "int"
.Fn route_msg_get_seq "struct route_msg *msg"
.Ft "const struct sockaddr *"
.Fn route_msg_get_dest "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_dest "struct route_msg *msg" "const struct sockaddr *dest"
.Ft "const struct sockaddr *"
.Fn route_msg_get_gateway "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_gateway "struct route_msg *msg" "const struct sockaddr *gateway"
.Ft "const struct sockaddr *"
.Fn route_msg_get_netmask "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_netmask "struct route_msg *msg" "const struct sockaddr *netmask"
.Ft "const struct sockaddr *"
.Fn route_msg_get_genmask "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_genmask "struct route_msg *msg" "const struct sockaddr *genmask"
.Ft "const struct sockaddr *"
.Fn route_msg_get_ifp "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_ifp "struct route_msg *msg" "const struct sockaddr *ifp"
.Ft "const struct sockaddr *"
.Fn route_msg_get_ifa "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_ifa "struct route_msg *msg" "const struct sockaddr *ifa"
.Ft "const struct sockaddr *"
.Fn route_msg_get_author "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_author "struct route_msg *msg" "const struct sockaddr *author"
.Ft "const struct sockaddr *"
.Fn route_msg_get_brd "struct route_msg *msg"
.Ft "int"
.Fn route_msg_set_brd "struct route_msg *msg" "const struct sockaddr *brd"
.Ft "int"
.Fn route_msg_decode "const u_char *data" "size_t dlen" "struct route_msg ***listp" "const char *mtype"
.Ft "int"
.Fn route_msg_encode "const struct route_msg *msg" "u_char *data" "size_t dlen"
.Ft "void"
.Fn route_msg_print "struct route_msg *msg" "FILE *fp"
.Ft "int"
.Fn route_msg_send "struct route_msg *msg" "int sock"
.Ft "int"
.Fn route_msg_recv "struct route_msg ***listp" "int sock" "const char *mtype"
.Sh DESCRIPTION
These functions support
.Nm "route_msg"
objects, used for communication over kernel routing sockets.
The routing socket messages are defined in the header file
.Li "<net/route.h>" .
Each
.Nm route_msg
instance is represented by a
.Li "struct route_msg" .
.Pp
.Fn route_msg_create
creates a new, empty
.Nm route_msg
object.
.Pp
.Fn route_msg_destroy
destroys the object pointed to by
.Fa "*msgp" .
Upon return,
.Fa "*msgp"
will be set to
.Dv NULL .
If
.Fa "*msgp"
is already
.Dv NULL
when
.Fn route_msg_destroy
is invoked, nothing happens.
.Pp
The
.Fn route_msg_get_XXX
and
.Fn route_msg_set_XXX
routines allow reading and changing the attributes associated with a
.Nm route_msg
object.
.Pp
.Fn route_msg_decode
decodes routing messages from data encoded in the format used when
communicating over kernel routing sockets.
The encoded data is pointed to by
.Fa data
and has length
.Fa dlen .
Zero or more
.Nm route_msg
objects are constructed and pointed to by the elements of a newly
allocated array of pointers.
The length of the array is returned by
.Fn route_msg_decode
and a pointer to the array is stored in
.Fa "*listp" .
The array itself is allocated with
.Xr typed_mem 3
type
.Fa mtype ,
and the caller is responsible for eventually freeing it (as well as
destroying the individudal
.Nm route_msg
objects).
If the encoded data is invalid, -1 is returned with
.Va errno
set to
.Er EINVAL .
.Pp
.Fn route_msg_encode
encodes a
.Nm route_msg
object into the format used when communicating over kernel routing sockets.
The resulting data is stored in the buffer pointed to by
.Fa data ,
which must have length at least
.Fa dlen .
If the encoding does not fit in
.Fa dlen
bytes, -1 is returned with
.Va errno
set to
.Er EMSGSIZE .
Otherwise, the number of bytes actually written into
.Fa data
is returned.
.Pp
.Fn route_msg_print
prints out a
.Nm route_msg
object to the stream
.Fa fp
for debugging purposes.
.Pp
.Fn route_msg_send
encodes and writes the
.Nm route_msg
object to the routing socket descriptor
.Fa sock .
.Pp
.Nm route_msg_recv
reads the next packet from the routing socket descriptor
.Fa sock
and decodes
.Nm route_msg
objects from the data read.
The
.Fa listp
and
.Fa mtype
parameters are the same as with
.Fn route_msg_decode .
.Sh RETURN VALUES
Except for functions returning
.Li "void" ,
all of the above functions return -1 or
.Dv NULL
to indicate an error, with
.Va errno
set appropriately.
.Sh SEE ALSO
.Xr if_util 3 ,
.Xr libpdel 3 ,
.Xr typed_mem 3 ,
.Xr uroute 3 ,
.Xr route 4
.Sh HISTORY
The PDEL library was developed at Packet Design, LLC.
.Dv "http://www.packetdesign.com/"
.Sh AUTHORS
.An Archie Cobbs Aq archie@freebsd.org


Librerie/libpdel-0.6.1/pdel/net/route_msg.c

/*
 * @COPYRIGHT@
 *
 * Author: Archie Cobbs <archie@freebsd.org>
 */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/socket.h>
#include <sys/sysctl.h>
#include <sys/sockio.h>
#include <sys/ioctl.h>

#include <net/if.h>
#include <net/if_dl.h>
#include <net/if_types.h>
#include <net/route.h>

#include <netinet/in.h>
#include <arpa/inet.h>

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <stdarg.h>
#include <unistd.h>
#include <string.h>
#include <assert.h>
#include <errno.h>
#include <time.h>

#include "structs/structs.h"
#include "structs/type/array.h"

#include "util/typed_mem.h"
#include "net/route_msg.h"

#define ROUNDUP(a) \
	((a) > 0 ? (1 + (((a) - 1) | (sizeof(long) - 1))) : sizeof(long))
#define ADVANCE(x, n)	((x) += ROUNDUP((n)->sa_len))

struct route_msg {
	int		type;		/* message type */
	int		index;		/* index for associated ifp */
	int		flags;		/* flags, incl. kern & msg, e.g. DONE */
	pid_t		pid;		/* for sender to identify action */
	int		seq;		/* for sender to identify action */
	int		error;		/* why failed */
	struct sockaddr	*dest;		/* destination sockaddr */
	struct sockaddr	*gateway;	/* gateway sockaddr */
	struct sockaddr	*netmask;	/* netmask sockaddr */
	struct sockaddr	*genmask;	/* genmask sockaddr */
	struct sockaddr	*ifp;		/* interface name sockaddr */
	struct sockaddr	*ifa;		/* interface addr sockaddr */
	struct sockaddr	*author;	/* author of redirect sockaddr */
	struct sockaddr	*brd;		/* for NEWADDR, broadcast or p2p dest */
};

/*
 * Internal variables
 */

/* List of sockaddr's in a route message from the kernel */
struct msg_sock {
	const char	*name;
	u_int		offset;
};

static const	struct msg_sock msg_socks[] = {
	{ "dest",	offsetof(struct route_msg, dest)	},
	{ "gateway",	offsetof(struct route_msg, gateway)	},
	{ "netmask",	offsetof(struct route_msg, netmask)	},
	{ "genmask",	offsetof(struct route_msg, genmask)	},
	{ "ifp",	offsetof(struct route_msg, ifp)		},
	{ "ifa",	offsetof(struct route_msg, ifa)		},
	{ "author",	offsetof(struct route_msg, author)	},
	{ "brd",	offsetof(struct route_msg, brd)		},
	{ NULL, 0 }
};

static int	prev_seq;		/* last used sequence number */

/*
 * Internal functions
 */
static int	route_msg_set_sa(struct route_msg *msg,
			const char *name, const struct sockaddr *sa);

/*
 * Create a new route message.
 */
struct route_msg *
route_msg_create(void)
{
	struct route_msg *msg;

	if ((msg = MALLOC("route_msg", sizeof(*msg))) == NULL)
		return (NULL);
	memset(msg, 0, sizeof(*msg));
	msg->seq = ++prev_seq;				/* XXX not atomic */
	msg->pid = getpid();
	return (msg);
}

/*
 * Destroy a routing message.
 */
void
route_msg_destroy(struct route_msg **msgp)
{
	struct route_msg *const msg = *msgp;

	if (msg == NULL)
		return;
	FREE("route_msg.dest", msg->dest);
	FREE("route_msg.gateway", msg->gateway);
	FREE("route_msg.netmask", msg->netmask);
	FREE("route_msg.genmask", msg->genmask);
	FREE("route_msg.ifp", msg->ifp);
	FREE("route_msg.ifa", msg->ifa);
	FREE("route_msg.author", msg->author);
	FREE("route_msg.brd", msg->brd);
	FREE("route_msg", msg);
	*msgp = NULL;
}

/*
 * Get message type.
 */
int
route_msg_get_type(struct route_msg *msg)
{
	return (msg->type);
}

/*
 * Set message type.
 */
void
route_msg_set_type(struct route_msg *msg, int type)
{
	msg->type = type;
}

/*
 * Get message index.
 */
int
route_msg_get_index(struct route_msg *msg)
{
	return (msg->index);
}

/*
 * Set message index.
 */
void
route_msg_set_index(struct route_msg *msg, int index)
{
	msg->index = index;
}

/*
 * Get message flags.
 */
int
route_msg_get_flags(struct route_msg *msg)
{
	return (msg->flags);
}

/*
 * Set message flags.
 */
void
route_msg_set_flags(struct route_msg *msg, int flags)
{
	msg->flags = flags;
}

/*
 * Get message error code.
 */
int
route_msg_get_error(struct route_msg *msg)
{
	return (msg->error);
}

/*
 * Get message process ID.
 */
pid_t
route_msg_get_pid(struct route_msg *msg)
{
	return (msg->pid);
}

/*
 * Get message sequence number.
 */
int
route_msg_get_seq(struct route_msg *msg)
{
	return (msg->seq);
}

/*
 * Get message destination.
 */
const struct sockaddr *
route_msg_get_dest(struct route_msg *msg)
{
	return (msg->dest);
}

/*
 * Set message destination.
 */
int
route_msg_set_dest(struct route_msg *msg, const struct sockaddr *dest)
{
	return (route_msg_set_sa(msg, "dest", dest));
}

/*
 * Get message gateway.
 */
const struct sockaddr *
route_msg_get_gateway(struct route_msg *msg)
{
	return (msg->gateway);
}

/*
 * Set message gateway.
 */
int
route_msg_set_gateway(struct route_msg *msg, const struct sockaddr *gateway)
{
	return (route_msg_set_sa(msg, "gateway", gateway));
}

/*
 * Get message netmask.
 */
const struct sockaddr *
route_msg_get_netmask(struct route_msg *msg)
{
	return (msg->netmask);
}

/*
 * Set message netmask.
 */
int
route_msg_set_netmask(struct route_msg *msg, const struct sockaddr *netmask)
{
	return (route_msg_set_sa(msg, "netmask", netmask));
}

/*
 * Get message cloning netmask.
 */
const struct sockaddr *
route_msg_get_genmask(struct route_msg *msg)
{
	return (msg->genmask);
}

/*
 * Set message cloning netmask.
 */
int
route_msg_set_genmask(struct route_msg *msg, const struct sockaddr *genmask)
{
	return (route_msg_set_sa(msg, "genmask", genmask));
}

/*
 * Get message interface name.
 */
const struct sockaddr *
route_msg_get_ifp(struct route_msg *msg)
{
	return (msg->ifp);
}

/*
 * Set message interface name.
 */
int
route_msg_set_ifp(struct route_msg *msg, const struct sockaddr *ifp)
{
	return (route_msg_set_sa(msg, "ifp", ifp));
}

/*
 * Get message interface address.
 */
const struct sockaddr *
route_msg_get_ifa(struct route_msg *msg)
{
	return (msg->ifa);
}

/*
 * Set message interface address.
 */
int
route_msg_set_ifa(struct route_msg *msg, const struct sockaddr *ifa)
{
	return (route_msg_set_sa(msg, "ifa", ifa));
}

/*
 * Get message author of the redirect message.
 */
const struct sockaddr *
route_msg_get_author(struct route_msg *msg)
{
	return (msg->author);
}

/*
 * Set message author of the redirect message.
 */
int
route_msg_set_author(struct route_msg *msg, const struct sockaddr *author)
{
	return (route_msg_set_sa(msg, "author", author));
}

/*
 * Get message broadcast or point-to-point destination address.
 */
const struct sockaddr *
route_msg_get_brd(struct route_msg *msg)
{
	return (msg->brd);
}

/*
 * Set message broadcast or point-to-point destination address.
 */
int
route_msg_set_brd(struct route_msg *msg, const struct sockaddr *brd)
{
	return (route_msg_set_sa(msg, "brd", brd));
}

/*
 * Set one of the struct sockaddr fields.
 */
static int
route_msg_set_sa(struct route_msg *msg,
	const char *name, const struct sockaddr *sa)
{
	struct sockaddr *copy;
	struct sockaddr **sp;
	char buf[32];
	int i;

	for (i = 0; msg_socks[i].name != NULL
	    && strcmp(name, msg_socks[i].name) != 0; i++);
	assert(msg_socks[i].name != NULL);
	sp = (struct sockaddr **)(void *)((u_char *)msg + msg_socks[i].offset);
	snprintf(buf, sizeof(buf), "route_msg.%s", name);
	if (sa != NULL) {
		if ((copy = MALLOC(buf, sa->sa_len)) == NULL)
			return (-1);
		memcpy(copy, sa, sa->sa_len);
	} else
		copy = NULL;
	FREE(buf, *sp);
	*sp = copy;
	return (0);
}

/*
 * Decode one or more routing messages from raw socket data.
 *
 * Returns the number of decoded messages, or -1 for error.
 * The array of struct route_msg pointers is returned in *listp,
 * in an array allocated with memory type 'mtype'.
 */
int
route_msg_decode(const u_char *data, size_t dlen,
	struct route_msg ***listp, const char *mtype)
{
	struct route_msg **list;
	struct rt_msghdr rtm;
	int posn;
	int num;
	int i;

	/* Count number of messages */
	for (num = posn = 0; posn < dlen; num++, posn += rtm.rtm_msglen) {
		if (dlen - posn < sizeof(rtm)) {
			errno = EINVAL;
			return (-1);
		}
		memcpy(&rtm, data + posn, sizeof(rtm.rtm_msglen));
		if (rtm.rtm_msglen < sizeof(rtm)) {
			errno = EINVAL;
			return (-1);
		}
	}

	/* Allocate list */
	if ((list = MALLOC(mtype, num * sizeof(*list))) == NULL)
		return (-1);
	memset(list, 0, num * sizeof(*list));

	/* Decode each message */
	for (i = posn = 0; i < num; i++, posn += rtm.rtm_msglen) {
		int posn2;
		int j;

		/* Create new route_msg object */
		if ((list[i] = route_msg_create()) == NULL)
			goto fail;

		/* Copy fields from route message header */
		memcpy(&rtm, data + posn, sizeof(rtm));
		list[i]->type = rtm.rtm_type;
		list[i]->index = rtm.rtm_index;
		list[i]->flags = rtm.rtm_flags;
		list[i]->pid = rtm.rtm_pid;
		list[i]->seq = rtm.rtm_seq;
		list[i]->error = rtm.rtm_errno;

		/* Create struct sockaddr's from each appended sockaddr */
		for (posn2 = posn + sizeof(rtm), j = 0;
		    msg_socks[j].name != NULL; j++) {
			const struct sockaddr *const sa
			    = (struct sockaddr *)(data + posn2);

			if ((rtm.rtm_addrs & (1 << j)) == 0)
				continue;
			if (sa->sa_len != 0) {
				if (route_msg_set_sa(list[i],
				    msg_socks[j].name, sa) == -1)
					goto fail;
			}
			ADVANCE(posn2, sa);
		}
	}

	/* Done */
	*listp = list;
	return (num);

fail:
	/* Cleanup after failure */
	for (i = 0; i < num; i++)
		route_msg_destroy(&list[i]);
	FREE(mtype, list);
	return (-1);
}

/*
 * Encode a route message into raw binary form.
 *
 * Returns the total encoded length, or -1 for error.
 */
int
route_msg_encode(const struct route_msg *msg, u_char *data, size_t dlen)
{
	struct rt_msghdr rtm;
	int totalLen;
	int i;

	/* Prepare header */
	if ((totalLen = sizeof(rtm)) > dlen) {
		errno = EMSGSIZE;
		return (-1);
	}
	memset(&rtm, 0, sizeof(rtm));
	rtm.rtm_version = RTM_VERSION;
	rtm.rtm_type = msg->type;
	rtm.rtm_index = msg->index;
	rtm.rtm_flags = msg->flags;
	rtm.rtm_pid = msg->pid;
	rtm.rtm_seq = msg->seq;
	rtm.rtm_errno = msg->error;

	/* Append struct sockaddr's */
	for (i = 0; msg_socks[i].name != NULL; i++) {
		struct sockaddr *const sa = *((struct sockaddr **)(void *)
		    ((u_char *)msg + msg_socks[i].offset));
		int len;

		if (sa == NULL)
			continue;
		rtm.rtm_addrs |= (1 << i);
		len = ROUNDUP(sa->sa_len);
		if (totalLen + len > dlen) {
			errno = EMSGSIZE;
			return (-1);
		}
		memcpy(data + totalLen, sa, sa->sa_len);
		memset(data + totalLen + sa->sa_len, 0, len - sa->sa_len);
		totalLen += len;
	}

	/* Copy in completed header */
	rtm.rtm_msglen = totalLen;
	memcpy(data, &rtm, sizeof(rtm));

	/* Done */
	return (totalLen);
}

/*
 * Send a route message to a socket.
 */
int
route_msg_send(struct route_msg *msg, int sock)
{
	u_char buf[512];
	int count;
	int elen;
	int r;

	if ((elen = route_msg_encode(msg, buf, sizeof(buf))) == -1)
		return (-1);
	for (count = 0; count < 32; count++) {
		if ((r = send(sock, buf, elen, 0)) >= 0)
			return (0);
		switch (errno) {
		case ENETUNREACH:
		case ESRCH:
			break;
		default:
			return (-1);
		}
	}
	return (-1);
}

/*
 * Receive route messages from a socket.
 */
int
route_msg_recv(struct route_msg ***listp, int sock, const char *mtype)
{
	u_char buf[512];
	int r;

	/* Read next raw message array */
	if ((r = recv(sock, buf, sizeof(buf), 0)) < 0)
		return (-1);

	/* Decode it */
	return (route_msg_decode(buf, r, listp, mtype));
}

/*
 * Print route message as a string.
 */
void
route_msg_print(struct route_msg *msg, FILE *fp)
{
	const char *typestr = NULL;

	fprintf(fp, "type=");
	switch (msg->type) {
	case RTM_ADD:		typestr = "ADD"; break;
	case RTM_DELETE:	typestr = "DELETE"; break;
	case RTM_CHANGE:	typestr = "CHANGE"; break;
	case RTM_GET:		typestr = "GET"; break;
	case RTM_LOSING:	typestr = "LOSING"; break;
	case RTM_REDIRECT:	typestr = "REDIRECT"; break;
	case RTM_MISS:		typestr = "MISS"; break;
	case RTM_LOCK:		typestr = "LOCK"; break;
	case RTM_OLDADD:	typestr = "OLDADD"; break;
	case RTM_OLDDEL:	typestr = "OLDDEL"; break;
	case RTM_RESOLVE:	typestr = "RESOLVE"; break;
	case RTM_NEWADDR:	typestr = "NEWADDR"; break;
	case RTM_DELADDR:	typestr = "DELADDR"; break;
	case RTM_IFINFO:	typestr = "IFINFO"; break;
	case RTM_NEWMADDR:	typestr = "NEWMADDR"; break;
	case RTM_DELMADDR:	typestr = "DELMADDR"; break;
	default:
		fprintf(fp, "%d", msg->type);
		break;
	}
	if (typestr != NULL)
		fprintf(fp, "%s", typestr);
	fprintf(fp, " index=%d flags=0x%x pid=%lu seq=%d errno=%d",
	    msg->index, msg->flags, (u_long)msg->pid, msg->seq, msg->error);
	if (msg->dest != NULL)
		fprintf(fp, " dest=%p", msg->dest);
	if (msg->gateway != NULL)
		fprintf(fp, " gateway=%p", msg->gateway);
	if (msg->netmask != NULL)
		fprintf(fp, " netmask=%p", msg->netmask);
	if (msg->genmask != NULL)
		fprintf(fp, " genmask=%p", msg->genmask);
	if (msg->ifp != NULL)
		fprintf(fp, " ifp=%p", msg->ifp);
	if (msg->ifa != NULL)
		fprintf(fp, " ifa=%p", msg->ifa);
	if (msg->author != NULL)
		fprintf(fp, " author=%p", msg->author);
	if (msg->brd != NULL)
		fprintf(fp, " brd=%p", msg->brd);
	fflush(fp);
}

